strawberryfields/circuitspecs/x.py

```diff
@@ -0,0 +1,256 @@
+        cov = (sf.hbar / 2) * S @ S.T
+
+        # Construct the A matrix
+        A = Amat(cov, hbar=sf.hbar)
+
+        # Construct the adjacency matrix represented by the A matrix.
+        # This must be an weighted, undirected bipartite graph. That is,
+        # B00 = B11 = 0 (no edges between the two vertex sets 0 and 1),
+        # and B01 == B10.T (undirected edges between the two vertex sets).
+        B = A[:n_modes, :n_modes]
+        B00 = B[:half_n_modes, :half_n_modes]
+        B01 = B[:half_n_modes, half_n_modes:]
+        B10 = B[half_n_modes:, :half_n_modes]
+        B11 = B[half_n_modes:, half_n_modes:]
+
+        # Perform unitary validation to ensure that the
+        # applied unitary is valid.
+
+        if not np.allclose(B00, 0) or not np.allclose(B11, 0):
+            # Not a bipartite graph
+            raise CircuitError(
+                "The applied unitary cannot mix between the modes {}-{} and modes {}-{}.".format(
+                    0, half_n_modes - 1, half_n_modes, n_modes - 1
+                )
+            )
+
+        if not np.allclose(B01, B10.T):
+            # Not a symmetric bipartite graph
+            raise CircuitError(
+                "The applied unitary on modes {}-{} must be identical to the applied unitary on modes {}-{}.".format(
+                    0, half_n_modes - 1, half_n_modes, n_modes - 1
+                )
+            )
+
+        # Now that the unitary has been validated, perform the Takagi decomposition
+        # to determine the constituent two-mode squeezing and interferometer
+        # parameters.
+        sqs, U = takagi(B01)
+        sqs = np.arctanh(sqs)
+
+        # ensure provided S2gates all have the allowed squeezing values
+        if not all(s in self.allowed_sq_ranges for s in sqs):
+            wrong_sq_values = [s for s in sqs if s not in self.allowed_sq_ranges]
+            raise CircuitError(
+                "Incorrect squeezing value(s) r={}. Allowed squeezing "
+                "value(s) are {}.".format(wrong_sq_values, self.allowed_sq_ranges)
+            )
+
+        # Convert the squeezing values into a sequence of S2gate commands
+        sq_seq = [
+            Command(ops.S2gate(sqs[i]), [registers[i], registers[i + half_n_modes]])
+            for i in range(half_n_modes)
+        ]
+
+        # NOTE: at some point, it might make sense to add a keyword argument to this method,
+        # to allow the user to specify if they want the interferometers decomposed or not.
+
+        # Convert the unitary into a sequence of MZgate and Rgate commands on the signal modes
+        U1 = ops.Interferometer(U, mesh="rectangular_symmetric", drop_identity=True)._decompose(
+            registers[:half_n_modes]
+        )
+
+        # Convert the unitary into a sequence of MZgate and Rgate commands on the idler modes
+        U2 = ops.Interferometer(U, mesh="rectangular_symmetric", drop_identity=True)._decompose(
+            registers[half_n_modes:]
+        )
```

Comment:
There is a potential optimization here: we could only perform the optimization _once_, and then shift the registers on the second decomposition:

```python
U1 = ops.Interferometer(U, mesh="rectangular_symmetric", drop_identity=True)._decompose(
    registers[:half_n_modes]
)

import copy
U2 = copy.deepcopy(U1)

for i in range(len(U2)):
    U2[i].reg = [registers[r.ind+half_n_modes] for r in U2[i].reg]
```
or something like that

Reply:
I like this idea very much!

Reply:
@nquesada, do you still want to implement this? 🙂

Reply:
Done :)